### PR TITLE
Allow shutdown if kill failed.

### LINKED ISSUE #1502: cylc stop --kill: failure to kill jobs prevent suite shut down
On `cylc stop --kill`, the suite will issue kill commands to active task jobs and then wait to shut down cleanly. Unfortunately, any failed kill commands (e.g. job host outage) will prevent the suite to shut down cleanly. This needs to be handled more elegantly. We may want to consider:
* Ignore failed kills and shut down any way?
* Retry kill commands with some delays?

### PR DESCRIPTION
**Pull request by arjclark**
Closes #1502 

If the kill part of shutdown --kill fails then a suite won't shutdown. This allows the suite to shutdown if the kill operation has failed.

@hjoliver - please review. I've been testing this by running a job on a remote machine and turning said machine off while the task was running then attempting to stop the suite. I'd imagine the same sort of testing would be possible if one were to disconnect from a network.

Can you give this a careful check over, I think this is right but am worried I may have missed one of the other shutdown combinations.

### CODE REVIEW

File: lib/cylc/scheduler.py [hunk 1]
```diff
@@ -1041,6 +1042,12 @@ def run(self):
                 proc_pool.close()
                 self.shut_down_now = True
 
+            if (self.shut_down_cleanly and self.pool.unkillable_only() and 
+                    self.kill_on_shutdown):
+                print '\nWARNING some tasks were not killable at shutdown'
```
Comment (hjoliver) [inline lib/cylc/scheduler.py]: Print to stderr.

**Comment by hjoliver** [comment]
@arjclark - admirable dedication to extreme testing, but you can just manually nuke the job with ```kill -9 PID``` :grin: 

If a job is unkillable because the suite has the wrong job ID or the task host is down, then this is good, but if it's unkillable because it suffered a hard kill (```kill -9```) then we should do a quick poll first - polling will detect that the job is already dead and allow the suite to shut down cleanly.

**Comment by hjoliver** [comment]
Another point - this will affect fast-moving test suites more than real suites, but it's not unusual to have tasks submit so close to the time that the kill command is issued, that they are not included in the list of tasks to be killed. If that happens, re-issuing the kill command usually works.  We should probably handle this better too.

**Comment by arjclark** [comment]
>admirable dedication to extreme testing ... unkillable because the suite has the wrong job ID or the task host is down

This is the main cause of `shutdown --kill` failure we see as our pool of linux servers is a bit up and down at times so was the best I could come up with for testing that :)

> it's not unusual to have tasks submit so close to the time that the kill command is issued, that they are not included in the list of tasks to be killed.

I'd thought invoking shutdown would block anything being moved to submitting? Will have a play and tweak accordingly.

> if kill fails, try a poll, and if that fails or reports that the task is still alive, then shut down anyway. Do you agree?

Sounds sensible to me, will add that in.

**Comment by arjclark** [comment]
@hjoliver - so, in current master it looks like a poll type operation is carried out anyway on issuing the kill command and the status gets updated if a task has been hard killed.

e.g. I run a suite, manually kill -9 the pid and issue a shutdown --kill, at that point the state of the task gets updated to failed:
```
2015-06-26T10:40:36+01 INFO - [foo.20100101T0000+01] -job(01) initiate job-kill
2015-06-26T10:40:36+01 INFO - Command succeeded: set_stop_cleanly(True)
2015-06-26T10:40:37+01 INFO - 0
2015-06-26T10:40:37+01 INFO - [foo.20100101T0000+01] -job(01) killed
2015-06-26T10:40:38+01 INFO - Suite shutting down at 2015-06-26T10:40:38+01
```

**Comment by matthewrmshin** [comment]
It is worth noting that a job kill no longer relies on information stored by the suite daemon. The `cylc job-kill` command uses only information in the `job.status` file.

A job kill should only fail if:
* Suite cannot connect to job host.
* Kill command not available for some reason.
* Job ID is no longer managed by the batch system, e.g. already killed.
  * The job has been killed by batch system in an untrappable way, e.g. out of memory.
  * The job was unable to send a message back to the suite daemon.

**Comment by matthewrmshin** [comment]
(The `TaskProxy.job_kill_callback` method makes some assumptions about a successful kill command as well.)

**Comment by hjoliver** [comment]
@arjclark - testing this on current master I don't get that behaviour:
```
2015-06-29T09:01:07+12 INFO - Command succeeded: set_stop_cleanly(True)
2015-06-29T09:01:14+12 INFO - 1
2015-06-29T09:01:14+12 WARNING - Traceback (most recent call last):
  File "/home/oliverh/cylc/cylc/bin/cylc-job-kill", line 40, in <module>
    main()
  File "/home/oliverh/cylc/cylc/bin/cylc-job-kill", line 36, in main
    sys.exit(BATCH_SYS_MANAGER.job_kill(parser.parse_args()[1][0]))
  File "/gpfs_hpcf/filesets/hpcf/home/oliverh/cylc/cylc.git/lib/cylc/batch_sys_manager.py", line 201, in job_kill
    os.killpg(int(pid), SIGKILL)
OSError: [Errno 3] No such process
2015-06-29T09:01:14+12 WARNING - [bar.2015-06-30T00] -job(01) kill failed
```
(and the suite does not shut down).

Re @matthewrmshin's comment:
> Job ID is no longer managed by the batch system, e.g. already killed
>  * The job has been killed by batch system in an untrappable way

In this case, the job has been killed by external means in an untrappable way (kill -9).  Manually invoking a poll results in a correct status update and subsequent shutdown.

(tested with local and remote 'background' and 'at' jobs).

**Comment by arjclark** [comment]
> testing this on current master I don't get that behaviour

Getting that error now myself. Must've had a floating uncommitted change when I swapped branch - d'oh!

**Comment by hjoliver** [comment]
Heh heh, I've done that before :grin:

**Comment by arjclark** [comment]
@hjoliver - Based on what's been said above re. when the kill mechanism can/can't work by Matt I've tweaked things slightly. 

Main change from above is that, if a "kill all" is issued, a poll is first issued to get things up to date.

>> it's not unusual to have tasks submit so close to the time that the kill command is issued, that they are not included in the list of tasks to be killed.

>I'd thought invoking shutdown would block anything being moved to submitting? Will have a play and tweak accordingly.

I'm still not sure how this can happen - is there a slight asynchronous problem happening here? I guess something needs fixing in the task_pool for shutdowns in general if this is possible? As far as I can tell the line `SuiteProcPool.get_inst().stop_job_submission()` should be stopping things progressing any further.

**Comment by hjoliver** [comment]
> I'm still not sure how this can happen ...

Sorry, I forgot to take another look at this yesterday.  I'm pretty sure that I've seen it happening, but I'm not sure of the reason - I'll try to reproduce it...

**Comment by hjoliver** [comment]
```
[cylc]
    cycle point format = %Y-%m-%d
[scheduling]
    initial cycle point = 2010-08-08
    [[dependencies]]
        [[[P1D]]]
            graph = foo => bar & baz => qux
```

If I run this simple dummy suite a few times, trying ```cylc stop --kill``` at some random point, it's not  hard to get unkilled tasks, e.g.:

kill-submit

Above, kill activity is highlighted in red, and unkilled task activity in blue.  (this is on master, btw).  It seems that submits initiated just before kills, but after the stop command is received, do not get killed.  Tasks that are already submitted or running at that point do get killed,  and submits initiated a bit later get skipped because the suite is stopping (yellow highlight).

**Comment by arjclark** [comment]
Ok, will see what I can do.

File: lib/cylc/scheduler.py [hunk 2]
```diff
@@ -917,7 +917,7 @@ def run(self):
             # Periodic check that the suite directory still exists
             # - designed to catch stalled suite daemons where the suite
             # directory has been deleted out from under itself
-            if datetime.datetime.now() > next_fs_check:
+            if datetime.datetime.utcnow() > next_fs_check:
```
Comment (arjclark) [inline lib/cylc/scheduler.py]: Snuck in a correction for time comparison in shutdown check for removed files.

**Comment by arjclark** [comment]
@hjoliver - I reckon this does the trick.

**Comment by hjoliver** [comment]
This looks good.  @matthewrmshin - reassigning for second review.

File: lib/cylc/scheduler.py [hunk 3]
```diff
@@ -388,9 +388,9 @@ def command_set_stop_cleanly(self, kill_active_tasks=False):
         """Stop job submission and set the flag for clean shutdown."""
         SuiteProcPool.get_inst().stop_job_submission()
         TaskProxy.stop_sim_mode_job_submission = True
-        if kill_active_tasks:
-            self.pool.kill_active_tasks()
         self.shut_down_cleanly = True
+        self.kill_on_shutdown = kill_active_tasks
+        self.next_kill_issue = datetime.datetime.utcnow()
```
Comment (matthewrmshin) [inline lib/cylc/scheduler.py]: I prefer `time.time` to `datetime.datetime.utcnow`. Although their run times are both insignificantly fast, the former is still about 10 times faster. It also makes it easier for the logic later on when you are adding 10 seconds, - you can just do `time.time() + 10.0` instead of have to wrap the delta with `datetime.timedelta`.

File: lib/cylc/task_pool.py [hunk 4]
```diff
@@ -643,6 +643,13 @@ def no_active_tasks(self):
                 return False
         return True
 
+    def unkillable_only(self):
```
Comment (matthewrmshin) [inline lib/cylc/task_pool.py]: What about `def has_unkillable_tasks`, which should make it instantly recognisable as a method returning a boolean?

Should also add doc string for a new method.

**Comment by matthewrmshin** [comment]
I guess this is quite tricky to write an automated test?

**Comment by arjclark** [comment]
> I guess this is quite tricky to write an automated test?

Yup, nigh on impossible so have to settle for it being tested as working by hand and not breaking anything else in the test battery.

Will make your suggested tweaks.

**Comment by arjclark** [comment]
@matthewrmshin - Using `time.time()` now, renamed `unkillable_only()` to `has_unkillable_tasks_only` and added a docstring as requested.

File: lib/cylc/scheduler.py (hunk 2 above)
Comment (matthewrmshin) [inline lib/cylc/scheduler.py]: I guess this is unrelated to this change?

Comment (arjclark) [inline lib/cylc/scheduler.py]: Yup, just a correction.

File: lib/cylc/scheduler.py (hunk 3 above)
Comment (matthewrmshin) [inline lib/cylc/scheduler.py]: (A gist to prove the claim above: https://gist.github.com/matthewrmshin/cabf70970b2dc7e85123)
